sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -2079,4 +2078,620 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
   override def third: Expression = toBase
 
   override def dataType: DataType = StringType
-}
\ No newline at end of file
+}
+
+case class GpuFormatNumber(x: Expression, d: Expression)
+    extends GpuBinaryExpression with ExpectsInputTypes with NullIntolerant {
+  
+  override def left: Expression = x
+  override def right: Expression = d
+  override def dataType: DataType = StringType
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, IntegerType)
+
+  private def reduceLeadingZeros(cv: ColumnVector): ColumnVector = {
+    // replace "[0]+" to "0" in a column vector
+    val (isZero, zeroCv) = withResource(Scalar.fromString("0")) { zero =>
+      (cv.startsWith(zero), ColumnVector.fromScalar(zero, cv.getRowCount.toInt))
+    }
+    withResource(isZero) { isZero =>
+      withResource(zeroCv) { zeroCv =>
+        isZero.ifElse(zeroCv, cv)
+      }
+    }
+  }
+
+  private def removeNegSign(cv: ColumnVector): ColumnVector = {
+    withResource(Scalar.fromString("-")) { negativeSign =>
+      cv.lstrip(negativeSign)
+    }
+  }
+
+  private def normalDoubleSplit(cv: ColumnVector, d: Int): (ColumnVector, ColumnVector) = {
+    val roundedStr = withResource(cv.round(d, RoundMode.HALF_EVEN)) { rounded =>
+      rounded.castTo(DType.STRING)
+    }
+    val (intPart, decPart) = withResource(roundedStr) { _ =>
+      withResource(roundedStr.stringSplit(".", 2)) { intAndDec =>
+        (intAndDec.getColumn(0).incRefCount(), intAndDec.getColumn(1).incRefCount())
+      }
+    }
+    val intPartNoNeg = withResource(intPart) { _ =>
+      removeNegSign(intPart)
+    }
+    val decPartPad = withResource(decPart) { _ =>
+      decPart.pad(d, PadSide.RIGHT, "0")
+    }
+    (intPartNoNeg, decPartPad)
+  }
+
+  private def expDoubleSplit(cv: ColumnVector, d: Int): (ColumnVector, ColumnVector) = {
+    // handle special case: 1.234567e+7 or 1.234567e-6
+    // get three parts first:
+    val replaceDelimToE = withResource(Scalar.fromString("e")) { e =>
+      withResource(Scalar.fromString(".")) { p =>
+        cv.stringReplace(e, p)
+      }
+    }
+    // get three parts: 1.234567e+7 -> 1, 234567, +7
+    val (intPartSign, decPart, expPart) = withResource(replaceDelimToE) { _ =>
+      withResource(replaceDelimToE.stringSplit(".", 3)) { intDecExp =>
+        (intDecExp.getColumn(0).incRefCount(),
```

Comment:
I don't think this works. I thought stringSplit returned an array, not a struct, so there is just one child column of an array.

Reply:
I think `stringSplit` returned a `Table` so we can getColumn from it. [cuDF doc](https://docs.rapids.ai/api/cudf-java/stable/ai/rapids/cudf/columnview#stringSplit-java.lang.String-int-)